app/src/main/java/org/mozilla/tv/firefox/MainActivity.kt

```diff
@@ -235,19 +234,46 @@ class MainActivity : LocaleAwareAppCompatActivity(), OnUrlEnteredListener, Media
 
     override fun dispatchKeyEvent(event: KeyEvent): Boolean {
         val fragmentManager = supportFragmentManager
-        val maybeBrowserFragment = (fragmentManager.findFragmentByTag(WebRenderFragment.FRAGMENT_TAG) as WebRenderFragment?)?.let {
+
+        val maybeBrowserFragment = (fragmentManager.findFragmentByTag(WebRenderFragment.FRAGMENT_TAG)
+                as WebRenderFragment?)?.let {
             if (it.isVisible) it else null
         }
-        val maybePocketFragment = (fragmentManager.findFragmentByTag(PocketVideoFragment.FRAGMENT_TAG) as PocketVideoFragment?)?.let {
-            if (it.isVisible) it else null
+
+        if (event.keyCode == KeyEvent.KEYCODE_MENU) {
+            if (event.action == KeyEvent.ACTION_DOWN) {
+                showOrHideOverlay()
+            }
+            return true
         }
+
         return videoVoiceCommandMediaSession.dispatchKeyEvent(event) ||
-                (maybePocketFragment?.dispatchKeyEvent(event) ?: false) ||
                 (maybeBrowserFragment?.dispatchKeyEvent(event) ?: false) ||
                 super.dispatchKeyEvent(event)
     }
 
-    companion object {
-        const val PARENT_FRAGMENT = "PARENT_FRAGMENT"
+    /**
+     * If overlay is shown, webRenderFragment will be shown
+     */
+    fun showOrHideOverlay() {
+        val maybeBrowserFragment = (supportFragmentManager.findFragmentByTag(WebRenderFragment.FRAGMENT_TAG)
+                as WebRenderFragment?)?.let {
+            if (it.isVisible) it else null
+        }
+        val maybePocketFragment = (supportFragmentManager.findFragmentByTag(PocketVideoFragment.FRAGMENT_TAG)
+                as PocketVideoFragment?)?.let {
+            if (it.isVisible) it else null
+        }
+        val overlayFragment = (supportFragmentManager.findFragmentByTag(NavigationOverlayFragment.FRAGMENT_TAG)
+                as NavigationOverlayFragment?)?.let {
+            if (it.isVisible) it else null
+        }
+
+        if (overlayFragment != null) {
+            serviceLocator.screenController.showBrowserScreenForCurrentSession(supportFragmentManager,
+                    webRenderComponents.sessionManager.selectedSessionOrThrow, true)
```

Comment:
I'm not sure that showing a particular page is really implied by 'hide overlay'.  I think it would simplify and decouple a lot of our logic (and make big UI changes easier) if we were able to add and remove an overlay fragment without destroying the browser underneath (that is, don't use `FragmentTransaction#replace`).  

Do you think that could be doable here, or is it out of scope?

Reply:
you mean to use `FragmentTrasnaction#add`? We aren't necessarily destroying the browser with `replace` (if you try debugging `showBrowserScreenForCurrentSession`, `WebRenderFragment` is not always destroyed (mainly because we are using `.addToBackStack(null)`)

Reply:
I agree with Severin's concern about coupling: we decide to show/hide the browserFragment based on overlay visibility so the browserFragment's visibility logic is now coupled to the overlay fragment.

To the user, however, the browserFragment is present all of the time: sometimes, it just has the overlay drawn on top of it. If the overlay shows/hides based on its own state and the browserFrag visibility does not need to respond accordingly, that state is decoupled.

We can write this a few ways:
- host the overlay as a subfragment in the BrowserFragment
- Add the overlay fragment into a separate viewGroup from  in MainActivity
- Add the overlay fragment into the same viewGroup as BrowserFragment in MainActivity (`fragmentTransaction.add`)
- ?

I like the first two (using the same viewGroup adds unnecessary depth and complexity) but otherwise don't have a preference: the former reintroduces some coupling (the overlay needs to attach itself to the view hierarchy) but could be better encapsulated. Though the overlay responds to global key events so maybe MainActivity should own it.